# TASK 1
## RECORD MATCHING
## Submitted by: Yashasvi Rajani
## Email: yashasvirajani302@gmail.com
## Github link: https://github.com/yashasvi01


### tool used: python
### IDE: colab

In [22]:
!pip install pandas

In [23]:
import pandas as pd

In [24]:
#read data from csv file into pandas dataframe
df1 = pd.read_csv("data1.csv")
df2 = pd.read_csv("data2.csv")

In [25]:
df1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [26]:
df2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

## PART I. TO IDENTIFY RECORDS PRESENT IN DATA1 BUT MSSING IN DATA2

In [27]:

primaryKey = ["Order ID", "Product ID"]

#PERFORMING outer join of the two datafframes
#indicator makes a column "_merge" WHICH SHOWS presence of elements
df_merged = pd.merge(df1, df2, on = primaryKey, how = "outer", indicator = True)

In [28]:
df_merged.tail(20)

Order ID       Product ID  Qty_x  Qty_y      _merge
9966  US-2017-116659  TEC-PH-10002824    NaN    3.0  right_only
9967  US-2017-117534  OFF-AR-10003087    NaN    9.0  right_only
9968  US-2017-117534  TEC-CO-10000971    NaN    2.0  right_only
9969  US-2017-118157  OFF-EN-10004459    NaN    2.0  right_only
9970  US-2017-129777  OFF-PA-10001970    NaN    2.0  right_only
9971  US-2017-130953  OFF-BI-10004828    NaN    2.0  right_only
9972  US-2017-133361  TEC-AC-10003280    NaN    3.0  right_only
9973  US-2017-133781  OFF-EN-10004483    NaN    3.0  right_only
9974  US-2017-141943  OFF-EN-10003448    NaN    2.0  right_only
9975  US-2017-147886  OFF-PA-10000232    NaN    2.0  right_only
9976  US-2017-148866  OFF-PA-10004782    NaN    5.0  right_only
9977  US-2017-149510  OFF-BI-10004233    NaN    6.0  right_only
9978  US-2017-152569  TEC-PH-10002185    NaN    2.0  right_only
9979  US-2017-158526  FUR-CH-10002602    NaN    8.0  right_only
9980  US-2017-160465  TEC-PH-10004522    NaN    6.0  right_only
9981  US-2017-165344  OFF-BI-10003196    NaN   10.0  right_only
9982  US-2017-165358  TEC-CO-10001943    NaN    5.0  right_only
9983  US-2017-167920  OFF-AP-10000159    NaN    5.0  right_only
9984  US-2017-169502  OFF-AP-10001947    NaN    5.0  right_only
9985  US-2017-169551  FUR-BO-10001519    NaN    3.0  right_only

In [29]:
missingRecords1 = df_merged[df_merged["_merge"] == "left_only"]
missingRecords1.head(10)

Order ID       Product ID  Qty_x  Qty_y     _merge
0    CA-2014-100006  TEC-PH-10002075    3.0    NaN  left_only
10   CA-2014-100678  OFF-EN-10000056    3.0    NaN  left_only
19   CA-2014-100895  OFF-AR-10004511    2.0    NaN  left_only
35   CA-2014-101560  OFF-BI-10000309    3.0    NaN  left_only
61   CA-2014-102673  OFF-LA-10001771   12.0    NaN  left_only
85   CA-2014-103366  TEC-AC-10003628    5.0    NaN  left_only
90   CA-2014-103429  OFF-PA-10001712    4.0    NaN  left_only
109  CA-2014-103940  OFF-ST-10000943    3.0    NaN  left_only
112  CA-2014-103989  FUR-FU-10003981    3.0    NaN  left_only
136  CA-2014-104738  TEC-PH-10000576    3.0    NaN  left_only

In [30]:
#ans1
len(missingRecords1)

507

### there are 507 records missing in data 2 which are present in data1


In [31]:
print(len(df1))
print(len(df2))

9478
9479


## PART II. TO IDENTIFY RECORDS PRESENT IN DATA2 BOT MSSING IN DATA1

In [32]:
#2nd part
missingRecords2 = df_merged[df_merged["_merge"] == "right_only"]
missingRecords2.head(10)

Order ID       Product ID  Qty_x  Qty_y      _merge
9478  CA-2014-100706  TEC-AC-10001314    NaN    2.0  right_only
9479  CA-2014-100762  OFF-PA-10001815    NaN    3.0  right_only
9480  CA-2014-101427  OFF-AR-10002257    NaN    3.0  right_only
9481  CA-2014-102652  FUR-FU-10001918    NaN    7.0  right_only
9482  CA-2014-102869  OFF-PA-10000788    NaN    3.0  right_only
9483  CA-2014-103310  TEC-PH-10001817    NaN    5.0  right_only
9484  CA-2014-103331  OFF-PA-10002160    NaN    3.0  right_only
9485  CA-2014-103492  OFF-BI-10004817    NaN    5.0  right_only
9486  CA-2014-104829  OFF-PA-10003016    NaN    7.0  right_only
9487  CA-2014-105165  FUR-TA-10004154    NaN    1.0  right_only

In [33]:
#ans2.
len(missingRecords2)

508

### there are 508 records missing in data1 which are present in data2


# PART III. sum of total Qty of records missing in data1 but present in data2

In [34]:
#related to previous solution
# we use the missing records table to calculate sum qty in right only
# marked as column "Qty_y"
sum(missingRecords2["Qty_y"])

1956.0

### sum of qty = 1956

## PART IV. TO FIND TOTAAL NUMBER OF UNIQUE RRECORDS:

### we use concat function for this, instead of merge
### because we need to stack the two datasets on each other
## inorder to find unquie records
## we do this using drop_duplicates() function

In [35]:
dfConcat = pd.concat([df1, df2])
dfConcat
# dfConcat is the concatenated dataframe

Order ID       Product ID  Qty
0     CA-2014-100006  TEC-PH-10002075    3
1     CA-2014-100090  FUR-TA-10003715    3
2     CA-2014-100090  OFF-BI-10001597    6
3     CA-2014-100293  OFF-PA-10000176    6
4     CA-2014-100328  OFF-BI-10000343    1
...              ...              ...  ...
9474  US-2017-169551  OFF-PA-10004100    3
9475  US-2017-169551  OFF-ST-10004835    3
9476  US-2017-169551  TEC-AC-10002018    3
9477  US-2017-169551  TEC-AC-10003033    2
9478  US-2017-169551  TEC-PH-10001363    2

[18957 rows x 3 columns]

In [36]:
len(dfConcat)

18957

In [37]:
len(dfConcat[primaryKey].drop_duplicates())

9986

therefor, total number of unique records is 9986